pkg/operator/storageclasscontroller/vmware.go

```diff
@@ -76,12 +77,31 @@ func NewStoragePolicyAPI(ctx context.Context, connection *vclib.VSphereConnectio
 	return storagePolicyAPIClient
 }
 
-func (v *storagePolicyAPI) GetDefaultDatastore(ctx context.Context) (*mo.Datastore, error) {
+func (v *storagePolicyAPI) GetDefaultDatastore(ctx context.Context, infra *v1.Infrastructure) (*mo.Datastore, error) {
 	vmClient := v.vcenterApiConnection.Client
 	config := v.vcenterApiConnection.Config
 	finder := find.NewFinder(vmClient.Client, false)
-	dcName := config.Workspace.Datacenter
-	dsName := config.Workspace.DefaultDatastore
+
+	// Following pattern of older generated ini config, default datastore is from FD[0], else if no FDs are define, we'll
+	// grab from deprecated fields
+	var dcName, dsName string
+	if infra.Spec.PlatformSpec.VSphere != nil && len(infra.Spec.PlatformSpec.VSphere.FailureDomains) > 0 {
+		// Due to first FD potentially not matching first vCenter in infra section, lets find correct one.
+		for _, fd := range infra.Spec.PlatformSpec.VSphere.FailureDomains {
+			if fd.Server == v.vcenterApiConnection.Hostname {
+				dcName = fd.Topology.Datacenter
+				dsName = fd.Topology.Datastore
+				break
+			}
+		}
+	} else {
+		if config != nil {
+			dcName = config.GetWorkspaceDatacenter()
+			dsName = config.GetDefaultDatastore()
+		} else {
+			return nil, fmt.Errorf("unable to determine default datastore from current config")
+		}
+	}
```

Comment:
So, this function is used only for populating migration datastore URL. Do we even need to read failure-domains for this one? Are workspace populated at all in multiple vcenter environment?

Reply:
When migrating to multi vCenter support, I am having the cluster now use the YAML styled new vsphere cloud provider config.  In this config, there is no concept of datastores.  In the 4.12 and older OCP clusters, no FD is defined and has workspace datastore populated.  Are you ok with this only populating when single vCenter is in use?

Also note that with multi vCenter, the migration function in the vsphere-syncer will log:
`{"level":"error","time":"2024-07-23T19:38:24.181570104Z","caller":"syncer/util.go:366","msg":"volume-migration feature is not supported on Multi-vCenter deployment","TraceId":"b8986369-a457-4670-a848-adb11353edd5","stacktrace":"sigs.k8s.io/vsphere-csi-driver/v3/pkg/syncer.initVolumeMigrationService\n\t/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/pkg/syncer/util.go:366\nsigs.k8s.io/vsphere-csi-driver/v3/pkg/syncer.podAdded\n\t/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/pkg/syncer/metadatasyncer.go:1851\nsigs.k8s.io/vsphere-csi-driver/v3/pkg/syncer.InitMetadataSyncer.func7\n\t/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/pkg/syncer/metadatasyncer.go:527\nk8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnAdd\n\t/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/vendor/k8s.io/client-go/tools/cache/controller.go:232\nk8s.io/client-go/tools/cache.(*processorListener).run.func1\n\t/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/vendor/k8s.io/client-go/tools/cache/shared_informer.go:909\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\t/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:157\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\t/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:158\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:135\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:92\nk8s.io/client-go/tools/cache.(*processorListener).run\n\t/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/vendor/k8s.io/client-go/tools/cache/shared_informer.go:903\nk8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1\n\t/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:75"}`

Basically setting the migration url will most likely get ignored.